### ISSUE TITLE
netty: Swap to UniformStreamByteDistributor

### ISSUE THREAD

**Issue by forthehell** [grpc/grpc-java#11659]
### Is your feature request related to a problem?

We send the  asyncUnaryCall in line in the same thread  and with only one channel  ， we find the rpc call arrived in server is out of order。

### Describe the solution you'd like

We find the reason is the WeightedFairQueueByteDistributor 。 And theUniformStreamByteDistributor  can more likely  do the job，  but there is no configuration to switch between those ByteDistributors。 can we add the configuration about it？ 

```java
if(FAIR_Distributor){
  WeightedFairQueueByteDistributor dist = new WeightedFairQueueByteDistributor(connection);
  dist.allocationQuantum(16 * 1024); // Make benchmarks fast again.
  streamByteDistributor = dist;
}else{
  UniformStreamByteDistributor uniformStreamByteDistributor =  new UniformStreamByteDistributor(connection);
  uniformStreamByteDistributor.minAllocationChunk(16 * 1024);
  streamByteDistributor = uniformStreamByteDistributor;
}
```

**Comment by ejona86** [comment]
If you start two asynchronous RPCs, there's no guarantees about their orders. The distributor doesn't even guarantee you what you're wanting, as the _message size_ can change the order.

Since [priorities were deprecated](https://httpwg.org/specs/rfc9113.html#PriorityHere) in RFC 9113, it might be reasonable to just change to UniformStreamByteDistributor unconditionally. We should also look into what's happening with Netty in this space (esp Netty 4.2), as they would have changed their default in the http2 builder, that grpc doesn't use.

So using the uniform distributor can be fair. But it is very bad to do it for your purpose. There are no ordering guarantees unless you get some response from the server before sending another RPC/message.

**Comment by forthehell** [comment]
> If you start two asynchronous RPCs, there's no guarantees about their orders. The distributor doesn't even guarantee you what you're wanting, as the _message size_ can change the order.
> 
> Since [priorities were deprecated](https://httpwg.org/specs/rfc9113.html#PriorityHere) in RFC 9113, it might be reasonable to just change to UniformStreamByteDistributor unconditionally. We should also look into what's happening with Netty in this space (esp Netty 4.2), as they would have changed their default in the http2 builder, that grpc doesn't use.
> 
> So using the uniform distributor can be fair. But it is very bad to do it for your purpose. There are no ordering guarantees unless you get some response from the server before sending another RPC/message.

I wonder why the _message size_ can change the order ?  asynchronous RPCs dataFrame will be sent to the WriterQueue  which belonged to the save channel。if flush the WriterQueue command ，they will be comsumed in order 。

**Comment by ejona86** [comment]
> I wonder why the message size can change the order ?

If the first message is bigger then the second message, then the fair distributor could finish sending the second message first. Whether this actually happens is dependent on the quantum size (16 KiB on that line), but that configuration is highly internal.

Even if they are _sent_ in-order, the server's executor could end up running the second one "first," due to both running concurrently (so which is "first" is dependent on your application logic) or kernel thread scheduling.

Then there's other cases like, "the two RPCs may not end up on the same connection," because of GOAWAY.

**Comment by ejona86** [comment]
This issue will look at swapping to UniformStreamByteDistributor, since priorities are deprecated. We should first understand what is happening in Netty, especially Netty 4.2. But our decision can be separate from theirs.

**Comment by shivaspeaks** [comment]
The history to note here is, we started with `WeightedFairQueueByteDistributor` which was also the default flow controller with netty. But then after a while we avoided the default and started using `UniformStreamByteDistributor` here, https://github.com/grpc/grpc-java/pull/1456 - some discussion happened in netty around headers could go out of context, https://github.com/netty/netty/issues/4758.
This change was undone somewhere in between and default was applied. I couldn't find that PR.

And then after that we have https://github.com/grpc/grpc-java/pull/3038 that talks about increasing message quantum with `WeightedFairQueueByteDistributor`. Why didn't we use Uniform flow control even when its QPS was higher (Ref: https://github.com/grpc/grpc-java/pull/3038#issuecomment-304960844)?

**Comment by kannanjgithub** [comment]
We can learn the history in due course, but the more important work to start with is to understand whether Netty's HTTP2 default builder has switched to the uniform distributor as mentioned in [this](https://github.com/grpc/grpc-java/issues/11659#issuecomment-2452672485) comment.

**Comment by shivaspeaks** [comment]
No, netty's default builder still uses [WeightedFairQueueByteDistributor](https://github.com/netty/netty/blob/4.2/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java#L64).

**Comment by ejona86** [comment]
The primary http2 builder in Netty is [Http2MultiplexCodecBuilder](https://github.com/netty/netty/blob/4.2/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodecBuilder.java). There's also [Http2FrameCodecBuilder](https://github.com/netty/netty/blob/4.2/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodecBuilder.java), which we would be more likely to use, but is rarely used by Netty users.

In those code flows, the [StreamByteDistribution is not being passed to the constructor](https://github.com/netty/netty/blob/1a59152cbaa7f6bce77bbc6de652d868fcb0ff17/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java#L54), so the constructor Shiva linked to is relevant.
